### PR TITLE
[fix](mysql-buffer) fix special buffer size with nested type

### PR DESCRIPTION
**Pull request by amorynan**
### What problem does this PR solve?
in special table with sepcial data insert into 
we could make sql with mysql return Malformed packet
```
mysql> select * from use1r;
ERROR 2027 (HY000): Malformed packet
```
this pr can resolve this problem
Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test 
    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [x] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

File: be/src/util/mysql_row_buffer.cpp [hunk 1]
```diff
@@ -107,7 +107,11 @@ MysqlRowBuffer<is_binary_format>::~MysqlRowBuffer() {
 template <bool is_binary_format>
 void MysqlRowBuffer<is_binary_format>::open_dynamic_mode() {
     if (!_dynamic_mode) {
-        *_pos++ = NEXT_EIGHT_BYTE;
+        // if _pos now exactly at the end of _buf memory,
+        // we should reserve 1 byte for _dynamic_mode flag byte to avoid *pos = 254
+        // cause _dynamic_mode flag byte be overwritten
+        reserve(1);
+        *_pos++ = NEXT_EIGHT_BYTE; // *_pos = 254 ; _pos++
```
Comment (xiaokang) [inline be/src/util/mysql_row_buffer.cpp:114]: Can you check that is there any more place should reserve() be called before *_pos++ ?

Comment (amorynan) [reply be/src/util/mysql_row_buffer.cpp:114]: I checked it, in be/src/util/mysql_row_buffer.cpp we have only one place to make *_pos++ 
and I also checked ```push_xxx``` func with reserve func or not , and anything else have I missed ？

**Review by eldenmoon — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Pull request merged**